### ISSUE TITLE
Can't build as static OR shared in Haiku (OS)

### ISSUE THREAD

**Issue by victroniko** [google/cpu_features#294]
Hello, I'm trying to build the [VOLK](https://github.com/gnuradio/volk) libraries on [Haiku](https://www.haiku-os.org/) (operating system). VOLK uses `cpu_features` as a submodule, I want/need to do a shared build but it doesn't work.

To isolate the problem I'm trying to build `cpu_features` alone. It seems no matter how I force it, explicitly using -DBUILD_SHARED_LIBS=ON and -DBUILD_PIC=ON args or even directly fumbling with those in CMakeLists.txt (ugly!) result is the same: 

> ...
> ld: CMakeFiles/list_cpu_features.dir/src/utils/list_cpu_features.c.o: relocation R_X86_64_PC32 against symbol 'gBumpAllocator' can not be used when making a shared object; recompile with -fPIC
> ld: final link failed: bad value
> collect2: error: ld returned 1 exit status
> CMakeFiles/list_cpu_features.dir/build.make:97: recipe for target 'list_cpu_features' failed

The reason to try a shared build is because the static one also doesn't build:

> ...
> ld: CMakeFiles/list_cpu_features.dir/src/utils/list_cpu_features.c.o: in function 'CreateTree':
> list_cpu_features.c:(.text+0x51d): undefined reference to 'GetX86Info'
> ld: list_cpu_features.c:(.text+0x529): undefined reference to 'GetX86CacheInfo'
> ld: list_cpu_features.c:(.text+0x531): undefined reference to 'FillX86BrandString'
> ld: list_cpu_features.c:(.text+0x617): undefined reference to 'GetX86Microarchitecture'
> ld: list_cpu_features.c:(.text+0x61e): undefined reference to 'GetX86MicroarchitectureName'
> ld: list_cpu_features.c:(.text+0x668): undefined reference to 'GetX86FeaturesEnumValue'
> ld: list_cpu_features.c:(.text+0x677): undefined reference to 'GetX86FeaturesEnumName'
> collect2: error: ld returned 1 exit status
> CMakeFiles/list_cpu_features.dir/build.make:97: recipe for target 'list_cpu_features' failed

In both cases, _list_cpu_features_ is the culprit. I'm stuck about how to properly approach this. Do I "disable" said command as recommended [here](https://github.com/google/cpu_features/issues/210#issuecomment-999356881), and build shared anyway? Or is there something to possibly do in `cpu_features` side that'll enable a correct static build?

FYI, version of things in Haiku are as follows: cmake 3.24.2-1, gcc 11.2.0, binutils 2.31.1-2

Any help will be appreciated :)

**Comment by toor1245** [comment]
@victroniko, actually we don't have any architecture support for Haiku OS., see https://github.com/google/cpu_features#whats-supported. Since for older processors we have to get the processor information from the OS, that's why you get errors like undefined reference to `DetectFeaturesFromOs` and `OverrideOsPreserves` even after the right build

I added test commit https://github.com/google/cpu_features/commit/7badfd0ab1cc66aa47c6e7b0cb029112219cd60f  and all works fine:
![image](https://user-images.githubusercontent.com/43444946/216771216-b4ed06ed-7d0c-40ff-995a-8165228a2b80.png)

tested on Haiku x86 32 bit via VirtualBox

probably for old processors we can get info via sysinfo:
![sysinfo_x86_64](https://user-images.githubusercontent.com/43444946/216771725-896787de-2070-40b4-97a6-6b172df17015.png)

also, in CMakeLists `CMAKE_SYSTEM_PROCESSOR MATCHE` it worth to add detection "x86" case:
![cpu_features_x86](https://user-images.githubusercontent.com/43444946/216771536-1e5f17bf-5511-4518-8b11-8ee5de8df2a6.png)

cc: @gchatelet, @Mizux

**Comment by victroniko** [comment]
Thank you for looking into this. Yes I know Haiku isn't supported... yet :)

The x86 variant is mantained mainly for binary compatibility with old BeOS 5.x, so things there are less than ideal for modern code. I was building on Haiku x86_64 (sorry I should've absolutely mentioned that!), as I believe it's the most sensible choice re: future support, codebase etc.

I tried anyway to build the test commit on x86_64, and it went further - _list_cpu_features_ now compiles fine. But as expected, final linking fails with a bunch of x86-specific undefined refs. 

![error1](https://user-images.githubusercontent.com/67852064/216794878-1de04d8f-ba2d-4a3f-b881-8301208f73d8.png)
...
![cpu_features-linkfail-haiku_x64](https://user-images.githubusercontent.com/67852064/216794741-10174cf5-d74b-498d-886a-4074ba8ee0da.png)

Below is a full build log from _haikuporter_ (Haiku's tool for building packages, similar to Gentoo _portage_).

[cpu_features-haikux64_buildlog_static.txt](https://github.com/google/cpu_features/files/10609919/cpu_features-haikux64_buildlog_static.txt)

**Comment by toor1245** [comment]
@victroniko, yes, it fails because I added test commit just to check list_cpu_features without testing.
You can run list_cpu_features with the following commands:
```
cmake -S. -Bcmake-build-release -DBUILD_TESTING=OFF -DCMAKE_BUILD_TYPE=Release
cmake --build cmake-build-release --config Release
./cmake-build-release/list_cpu_features
```

About x86, no matter what architecture x86 or x86_64 is the same code in cpu_features. I didn't use BeOS compatibility using gcc 2.95 (need to change code to support the compiler), I used the `setarch x86` command which allows you to use the latest version of gcc.

I added haiku_x86 impl file to tests and disabled OS specific tests (https://github.com/google/cpu_features/commit/484a95133ca6493d80cf76782c9d15de9e1aa0d2) and it works 

![image](https://user-images.githubusercontent.com/43444946/216817463-3052caab-404c-476d-a106-b89442d77cd7.png)

**Comment by victroniko** [comment]
I can confirm it now builds (as stand-alone) correctly on Haiku x86_64. Still requires a bit more massaging to install succesfully. This is what happens if we follow install [instructions](https://github.com/google/cpu_features#build-with-cmake) as-is:

> ...
Install the project...
/boot/system/bin/cmake -P cmake_install.cmake
-- Install configuration: "RelWithDebInfo"
-- Installing: /boot/system/lib/libcpu_features.a
CMake Error at cmake_install.cmake:41 (file):
  file INSTALL cannot copy file
  "/sources/cpu_features-484a95133ca6493d80cf76782c9d15de9e1aa0d2/build/libcpu_features.a"
  to "/boot/system/lib/libcpu_features.a": Read-only file system.
Makefile:109: recipe for target 'install' failed

This doesn't work because Haiku system-dirs are read only: installed packages are mounted on boot then "overlayed" on top of FS. So at install step we have to use a [path-abstraction variable](https://github.com/haikuports/haikuports/wiki/HaikuPorter-BuildRecipes#input) along with `DESTDIR=`, **in this way cpu_features builds to 100%**. But it seems it wants to install to _absolute_ paths on /boot - there is NO abstraction for that, only for libs bin devel docs... etc.

To make a crude example, this is the resulting package built with `DESTDIR=$libsDir` (/boot/system/lib), and the final path where things end:

![cpu_features-pkg](https://user-images.githubusercontent.com/67852064/216865171-3421f17b-003b-43b3-853d-a2017695cc05.png)
![cpu_features-wrongpath](https://user-images.githubusercontent.com/67852064/216865745-f94e5b40-2b26-4f88-b8ec-fd6647896a31.png)

As can be seen everything after `$libsDir` is hardcoded to /boot. With `DESTDIR=$binDir` things end in /boot/system/bin/boot/*, and so on. **Anyway it _does_ install**, and list_cpu_features can be ran by manually cd'ing to resulting path. So it seems this is the last thing before a fully working standalone build.

Now I'm off to test if it builds as submodule of VOLK...

**Comment by toor1245** [comment]
@victroniko, according to the haiku cli documentation and os discussion, we can install them to `~/config/non-packaged/bin/` and this folder is user writable.

refs:
https://www.haiku-os.org/docs/userguide/en/applications/cli-apps.html
https://www.haiku-os.org/docs/userguide/en/filesystem-layout.html
https://discuss.haiku-os.org/t/how-do-i-make-a-libary-for-haiku-os/7075/4

Added install path https://github.com/google/cpu_features/commit/674b5a471adb8b480e527169482d330f632ec6c3

```cmake
if(CMAKE_SYSTEM_NAME MATCHES "Haiku")
  set(CMAKE_INSTALL_LIBDIR ~/config/non-packaged/lib)
  set(CMAKE_INSTALL_BINDIR ~/config/non-packaged/bin)
  set(CMAKE_INSTALL_INCLUDEDIR ~/config/non-packaged/include)
endif()
```

![image](https://user-images.githubusercontent.com/43444946/217383668-9aeb3a5f-2dd0-46c8-83ca-53877555d91c.png)

cc: @Mizux

**Comment by toor1245** [comment]
@gchatelet, what do you think, can we try to add support Haiku OS?

**Comment by gchatelet** [comment]
I was about to reply. It would be nice to support Haiku OS but only if we can have a CI machine. @Mizux ?
